electrumsv/network.py

```diff
@@ -1027,305 +290,144 @@ async def _main_task(self) -> None:
 
-        proxy = config.get('proxy', None)
-        if isinstance(proxy, str):
-            proxy = SVProxy.from_string(proxy)
+    async def get_esv_server_tip_height(self) -> int:
+        """raises `ServiceUnavailableError`"""
+        if self.esv_client_cached_tip:
+            return cast(int, self.esv_client_cached_tip.height)
 
-        logger.info("main server: %s, proxy: %s", main_server, proxy)
-        return main_server, proxy
+        assert self.esv_client is not None
+        tip = await self.esv_client.get_chain_tips(longest_chain_only=True)
+        self.esv_client_cached_tip = chain_tip_to_header_obj(tip)
+        logger.debug(f"Fetched chain tip: {self.esv_client_cached_tip}")
+        assert self.esv_client_cached_tip is not None
+        return cast(int, self.esv_client_cached_tip.height)
 
-    def _available_servers(self, protocol: str) -> List[SVServer]:
-        now = time.time()
-        unchosen = set(SVServer.all_servers.values()).difference(self._chosen_servers)
-        return [server for server in unchosen
-                if server.protocol == protocol and server.state.can_retry(now)]
 
-    def _random_server_nowait(self, protocol: str) -> Optional[SVServer]:
-        servers = self._available_servers(protocol)
-        return random.choice(servers) if servers else None
+    def get_esv_server_tip_hash(self) -> Optional[bytes]:
+        if not self.esv_client_cached_tip:
+            return None
+        assert self.esv_client_cached_tip is not None
+        return cast(bytes, self.esv_client_cached_tip.hash)
 
-    async def _random_server(self, protocol: str) -> SVServer:
+    # TODO(post-1.4.0) Need a sound system for banning and selecting a better provider
+    async def _initial_headers_sync(self) -> None:
         while True:
-            server = self._random_server_nowait(protocol)
-            if server:
-                return server
-            await sleep(10)
-
-    async def _main_session(self) -> SVSession:
+            try:
+                while self.get_local_height() < await self.get_esv_server_tip_height():
+                    missing_heights = [height for height in range(
+                        self.get_local_height(), await self.get_esv_server_tip_height() + 1)]
+                    await self._fetch_missing_headers_at_heights(missing_heights)
+
+                assert self.esv_client_cached_tip is not None
+                if self.get_local_tip_hash() != self.get_esv_server_tip_hash():
+                    # Wallet can still operate like this so log error and return from function
+                    # TODO(1.4.0) Switch to alternative service providers
+                    logger.error("The Headers API tip is either lagging our previously "
+                        "stored tip or is an equal height fork. Local tip height: %s, "
+                        "Server tip height: %s", self.get_local_height(),
+                        self.esv_client_cached_tip.height)
```

Comment:
There is no such thing as an "esv client cached tip" this means nothing meaningful. Use a directly relevant name.